config/mysqlpersister/mysqlpersister.go

```diff
@@ -0,0 +1,198 @@
+package mysqlpersister
+
+import (
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"sort"
+	"sync"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/square/quotaservice/config"
+	"github.com/square/quotaservice/logging"
+	qsc "github.com/square/quotaservice/protos/config"
+)
+
+const (
+	pollingInterval = 1 * time.Second
+)
+
+type mysqlPersister struct {
+	latestVersion int
+	db            *sqlx.DB
+	m             *sync.RWMutex
+
+	watcher  chan struct{}
+	shutdown chan struct{}
+
+	activeFetchers *sync.WaitGroup
+
+	configs map[int]*qsc.ServiceConfig
+}
+
+type configRow struct {
+	Version int    `db:"Version"`
+	Config  string `db:"Config"`
+}
+
+func New(dbUser, dbPass, dbHost string, dbPort int, dbName string) (config.ConfigPersister, error) {
+	db, err := sqlx.Open("mysql",
+		fmt.Sprintf("%s:%s@(%s:%v)/%s",
+			dbUser,
+			dbPass,
+			dbHost,
+			dbPort,
+			dbName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Query("SELECT 1 FROM quotaservice LIMIT 1")
+	if err != nil {
+		return nil, errors.New("table quotaservice does not exist")
+	}
+
+	mp := &mysqlPersister{
+		db:             db,
+		configs:        make(map[int]*qsc.ServiceConfig),
+		activeFetchers: &sync.WaitGroup{},
+		m:              &sync.RWMutex{},
+		watcher:        make(chan struct{}),
+		shutdown:       make(chan struct{}),
+		latestVersion:  -1,
+	}
+
+	mp.pullConfigs()
+
+	mp.activeFetchers.Add(1)
+	go mp.configFetcher()
+
+	return mp, nil
+}
+
+func (mp *mysqlPersister) configFetcher() {
+	defer mp.activeFetchers.Done()
+
+	for {
+		select {
+		case <-time.After(pollingInterval):
+			if mp.pullConfigs() {
+				mp.notifyWatcher()
+			}
+		case <-mp.shutdown:
+			logging.Print("Received shutdown signal, shutting down mysql watcher")
+			return
+		}
+	}
+}
+
+// pullConfigs checks the database for new configs and returns true if there is a new config
+func (mp *mysqlPersister) pullConfigs() bool {
+	mp.m.RLock()
+	v := mp.latestVersion
+	mp.m.RUnlock()
+
+	var rows []*configRow
+	err := mp.db.Select(&rows, "SELECT Version, Config FROM quotaservice WHERE Version > ? ORDER BY Version ASC", v)
```

Comment:
I think using a magic string for the query is more error prone than building the query programmatically.

I'm not familiar with sqlx, but github.com/square/squalor has functionality like
```
q := books.Select(books.All()).Where(books.C("author").Eq(1234))
```

Reply:
Sqlx doesn't have a query builder, but I can't use squalor as I'm passing a blob with possible non utf-8 characters (https://github.com/square/squalor/blob/master/encode.go#L103-L109)

Reply:
I converted it to use [Squirrel](https://github.com/Masterminds/squirrel) which builds the query and then lets sqlx handle the querying.